File: plum-frontend/src/dto/Filter.ts
export enum FilterDimension {
    Category = 'category',
    Uri = 'uri',
    Status = 'status',
    Referer = 'referer',
    UserAgent = 'userAgent',
}

export type Filter = {
    [dimension in FilterDimension]?: string;
};

export const FilterLabels: { [dimension in FilterDimension]: string } = {
    [FilterDimension.Category]: 'Category',
    [FilterDimension.Uri]: 'Page',
    [FilterDimension.Status]: 'Status',
    [FilterDimension.Referer]: 'Referer',
    [FilterDimension.UserAgent]: 'User agent',
};

export function filtersEqual(a: Filter, b: Filter): boolean {
    return Object.values(FilterDimension).every(dimension => a[dimension] === b[dimension]);
}

File: core/filter.go
package core

type Filter struct {
	Category  Category
	Uri       string
	Status    string
	Referer   string
	UserAgent string
}

func (f Filter) MatchesUserAgent(userAgent UserAgent) bool {
	return f.UserAgent == "" || f.UserAgent == userAgent.Name()
}

func (f Filter) MatchesCategory(category Category) bool {
	return f.Category.IsZero() || f.Category == category
}

func (f Filter) MatchesUri(uri string) bool {
	return f.Uri == "" || f.Uri == uri
}

func (f Filter) MatchesStatus(status string) bool {
	return f.Status == "" || f.Status == status
}

func (f Filter) MatchesReferer(referer string) bool {
	return f.Referer == "" || f.Referer == referer
}
